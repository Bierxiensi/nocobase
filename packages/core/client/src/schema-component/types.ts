import type { Form } from '@formily/core';
import type { IRecursionFieldProps, ISchemaFieldProps, SchemaReactComponents } from '@formily/react';
import type React from 'react';

export interface ISchemaComponentContext {
  scope?: any;
  components?: SchemaReactComponents;
  refresh?: () => void;
  reset?: () => void;
  designable?: boolean;
  setDesignable?: (value: boolean) => void;
  SchemaField?: React.FC<ISchemaFieldProps>;
}

export interface ISchemaComponentProvider {
  designable?: boolean;
  form?: Form;
  scope?: any;
  components?: SchemaReactComponents;
}

export interface IRecursionComponentProps extends IRecursionFieldProps {
  scope?: any;
  components?: SchemaReactComponents;
}

export interface ISchemaComponentOptionsProps {
  scope?: any;
  components?: SchemaReactComponents;
  inherit?: boolean;
}
