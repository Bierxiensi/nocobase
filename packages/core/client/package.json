{
  "name": "@nocobase/client",
  "version": "0.9.3-alpha.1",
  "license": "Apache-2.0",
  "main": "lib",
  "module": "es/index.js",
  "typings": "es/index.d.ts",
  "dependencies": {
    "@antv/g2plot": "^2.4.31",
    "@dnd-kit/core": "^5.0.3",
    "@dnd-kit/sortable": "^6.0.1",
    "@emotion/css": "^11.11.0",
    "@formily/antd": "2.0.20",
    "@formily/core": "2.0.20",
    "@formily/react": "2.0.20",
    "@nocobase/evaluators": "0.9.3-alpha.1",
    "@nocobase/sdk": "0.9.3-alpha.1",
    "@nocobase/utils": "0.9.3-alpha.1",
    "ahooks": "^3.7.7",
    "antd": "^4.24.10",
    "axios": "^0.26.1",
    "classnames": "^2.3.2",
    "cron-parser": "^4.8.1",
    "cronstrue": "^2.27.0",
    "file-saver": "^2.0.5",
    "i18next": "^22.4.15",
    "i18next-http-backend": "^2.2.0",
    "markdown-it": "13.0.1",
    "markdown-it-highlightjs": "3.3.1",
    "mathjs": "^10.6.4",
    "mermaid": "9.4.3",
    "react-beautiful-dnd": "^13.1.1",
    "react-big-calendar": "^0.38.9",
    "react-drag-listview": "^0.1.9",
    "react-error-boundary": "^4.0.4",
    "react-helmet": "^6.1.0",
    "react-hotkeys-hook": "^3.4.7",
    "react-i18next": "^11.18.6",
    "react-iframe": "~1.8.5",
    "react-image-lightbox": "^5.1.4",
    "react-js-cron": "^3.2.0",
    "react-quill": "^1.3.5",
    "react-router-dom": "^5.3.4",
    "react-to-print": "^2.14.12",
    "sanitize-html": "2.10.0",
    "solarlunar-es": "^1.0.9",
    "use-deep-compare-effect": "^1.8.1"
  },
  "peerDependencies": {
    "@types/react": ">=16.8.0 || >=17.0.0",
    "@types/react-dom": ">=16.8.0 || >=17.0.0",
    "react": ">=16.8.0 || >=17.0.0",
    "react-dom": ">=16.8.0",
    "react-is": ">=16.8.0 || >=17.0.0"
  },
  "devDependencies": {
    "@testing-library/react": "^12.1.5",
    "@types/markdown-it": "12.2.3",
    "@types/markdown-it-highlightjs": "3.3.1",
    "axios-mock-adapter": "^1.21.4"
  },
  "gitHead": "ce588eefb0bfc50f7d5bbee575e0b5e843bf6644"
}
