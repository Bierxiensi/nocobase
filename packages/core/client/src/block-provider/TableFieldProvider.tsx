import { ArrayField, Field } from '@formily/core';
import { useField, useFieldSchema } from '@formily/react';
import React, { createContext, useContext, useEffect } from 'react';
import { APIClient } from '../api-client';
import { useCollectionField, useCollectionManager } from '../collection-manager';
import { useRecord } from '../record-provider';
import { BlockProvider, useBlockRequestContext } from './BlockProvider';
import { useFormBlockContext } from './FormBlockProvider';
import { useFormFieldContext } from './FormFieldProvider';
import { IsTreeTableContext, TableBlockContext, useAssociationNames } from './TableBlockProvider';

export const TableFieldContext = createContext<any>({});

const InternalTableFieldProvider = (props) => {
  const { params = {}, showIndex, dragSort, fieldName, rowKey = 'id' } = props;
  const field = useField();
  const fieldSchema = useFieldSchema();
  const { resource, service } = useBlockRequestContext();
  const treeTable = useContext(IsTreeTableContext);

  const formBlockCtx = useFormBlockContext();
  const formFieldCtx = useFormFieldContext();

  const fullFieldName = formFieldCtx && formFieldCtx.fieldName ? `${formFieldCtx.fieldName}.${fieldName}` : fieldName;

  if (!formBlockCtx?.updateAssociationValues?.includes(fullFieldName)) {
    formBlockCtx?.updateAssociationValues?.push(fullFieldName);
  }
  // if (service.loading) {
  //   return <Spin />;
  // }
  return (
    <TableFieldContext.Provider
      value={{
        field,
        fieldSchema,
        service,
        resource,
        params,
        showIndex,
        dragSort,
        rowKey: treeTable ? '__index' : rowKey,
        idKey: rowKey,
        treeTable: useIsEnableTree(props),
      }}
    >
      {props.children}
    </TableFieldContext.Provider>
  );
};

export const flat = (tree: any[]) => {
  return tree.reduce((pre, cur) => {
    return pre.concat(cur, flat(cur.children ?? []));
  }, []);
};

export const toTree = (adjacencyList) => {
  const buildSubtree = (node) => {
    const parentIndex = node.__index;
    return {
      ...node,
      children: adjacencyList
        .filter(
          (i) =>
            i.__index.length > parentIndex.length &&
            i.__index.startsWith(parentIndex) &&
            !i.__index.replace(`${parentIndex}.children.`, '').includes('.'),
        )
        .map(buildSubtree),
    };
  };
  return adjacencyList.filter((i) => !i.__index.includes('.')).map(buildSubtree);
};

export class TableFieldResource {
  field: Field;
  api: APIClient;
  sourceId: any;
  resource?: any;

  constructor(options) {
    this.field = options.field;
    this.api = options.api;
    this.sourceId = options.sourceId;
    this.resource = this.api.resource(options.resource, this.sourceId);
  }

  async list(options) {
    this.field.data = this.field.data || {};
    if (this.field.data.changed) {
      console.log('list.dataSource', this.field.data.dataSource);
      return {
        data: {
          data: options.tree ? toTree(this.field.data.dataSource) : this.field.data.dataSource,
        },
      };
    }
    if (!this.sourceId) {
      console.log('list.sourceId', this.field.data.dataSource);
      this.field.data.dataSource = [];
      return {
        data: {
          data: [],
        },
      };
    }
    const response = await this.resource.list(options);
    console.log('list', response);
    this.field.data.dataSource = options.tree ? flat(response.data.data) : response.data.data;
    return {
      data: {
        data: response.data.data,
      },
    };
  }

  async get(options) {
    console.log('get', options);
    const { filterByTk, treeTable } = options;
    return {
      data: {
        data: treeTable
          ? this.field.data.dataSource.find((i) => i.__index == filterByTk)
          : this.field.data.dataSource[filterByTk],
      },
    };
  }

  async create(options) {
    console.log('create', options);
    const { values, treeTable } = options;
    if (treeTable) {
      const tops = this.field.data.dataSource.filter((i) => !i.__index.includes('.')).length;
      this.field.data.dataSource = this.field.data.dataSource.concat({
        ...values,
        __index:
          (values.parent?.__parent && Object.keys(values.parent?.__parent).length === 0) || !values.parent?.__index
            ? String(tops)
            : `${values.parent?.__index}.children.${values.parent.children?.length ?? 0}`,
      });
    } else {
      this.field.data.dataSource = this.field.data.dataSource.concat(values);
    }
    this.field.data.changed = true;
  }

  async update(options) {
    console.log('update', options);
    const { filterByTk, values, treeTable } = options;

    if (treeTable) {
      const parentId = values.parent.id;
      const parentIndex = this.field.data.dataSource.find((i) => i.id === parentId).__index;
      const parentChildrenLen = (values.parent.children ?? []).length;

      values.__index = `${parentIndex}.children.${parentChildrenLen}`;

      const writeChildrenIndex = (node) => {
        (node.children ?? []).forEach((i, index) => {
          i.__index = `${node.__index}.children.${index}`;
          writeChildrenIndex(i);
        });
      };

      writeChildrenIndex(values);

      this.field.data.dataSource = this.field.data.dataSource.map((i, index) =>
        treeTable && filterByTk === i.__index ? values : index === filterByTk ? values : i,
      );
    } else {
      this.field.data.dataSource = this.field.data.dataSource.map((i, index) => (index === filterByTk ? values : i));
    }
    this.field.data.changed = true;
  }

  async destroy(options) {
    console.log('destroy', options);
    let { filterByTk, idKey, treeTable } = options;
    if (!Array.isArray(filterByTk)) {
      filterByTk = [filterByTk];
    }
    this.field.data.dataSource = this.field.data.dataSource.filter((item, index) => {
      return !filterByTk.includes(treeTable && idKey ? item[idKey] ?? item.__index : index);
    });
    this.field.data.changed = true;
  }
}

export const WithoutTableFieldResource = createContext(null);

export const TableFieldProvider = (props) => {
  const params = { ...props.params };
  if (useIsEnableTree(props)) {
    params.tree = true;
    params.filter = {
      ...(params.filter ?? {}),
      parentId: useRecord()?.id ?? null,
    };
    params.appends = params.appends ?? useAssociationNames(props.collection).filter((i) => i !== 'children');
  }
  return (
    <WithoutTableFieldResource.Provider value={false}>
      <BlockProvider block={'TableField'} {...props} params={params}>
        <InternalTableFieldProvider {...props} />
      </BlockProvider>
    </WithoutTableFieldResource.Provider>
  );
};

export const useTableFieldContext = () => {
  return useContext(TableFieldContext);
};

export const useTableFieldProps = () => {
  const field = useField<ArrayField>();
  const ctx = useTableFieldContext();
  useEffect(() => {
    if (!ctx?.service?.loading) {
      field.value = ctx?.service?.data?.data;
      field.data = field.data || {};
      field.data.selectedRowKeys = ctx?.field?.data?.selectedRowKeys;
    }
  }, [ctx?.service?.loading]);
  return {
    ...ctx,
    size: 'middle',
    loading: ctx?.service?.loading,
    showIndex: ctx.showIndex,
    dragSort: ctx.dragSort,
    pagination: false,
    required: ctx?.fieldSchema?.parent?.required,
    onRowSelectionChange(selectedRowKeys) {
      ctx.field.data = ctx?.field?.data || {};
      ctx.field.data.selectedRowKeys = selectedRowKeys;
    },
    onChange({ current, pageSize }) {
      ctx.service.run({ page: current, pageSize });
    },
  };
};

const useIsEnableTree = (props) => {
  const { getCollection } = useCollectionManager();

  return getCollection(props.collection).tree === 'adjacencyList';
};
