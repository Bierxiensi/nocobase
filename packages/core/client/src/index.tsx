import 'antd/dist/antd.css';
import './global.less';
export * as reactI18nNext from 'react-i18next';
export * as antd from 'antd';
export * as formilyReact from '@formily/react';
export * as antdIcons from '@ant-design/icons';
export * as emotionCss from '@emotion/css';

export * from './acl';
export * from './antd-config-provider';
export * from './api-client';
export * from './application';
export * from './async-data-provider';
export * from './block-provider';
export * from './board';
export * from './china-region';
export * from './collection-manager';
export * from './document-title';
export * from './i18n';
export * from './icon';
export * from './plugin-manager';
export * from './pm';
export * from './powered-by';
export * from './record-provider';
export * from './route-switch';
export * from './schema-component';
export * from './schema-initializer';
export * from './schema-settings';
export * from './schema-templates';
export * from './settings-form';
export * from './system-settings';
export * from './user';
