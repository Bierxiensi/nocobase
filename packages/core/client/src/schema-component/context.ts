import { createContext } from 'react';
import type { ISchemaComponentContext } from './types';

export const SchemaComponentContext = createContext<ISchemaComponentContext>({});
