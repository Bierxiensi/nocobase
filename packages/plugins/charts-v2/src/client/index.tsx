import {
  SchemaComponentOptions,
  SchemaInitializerContext,
  SchemaInitializerProvider,
  ACLCollectionProvider,
} from '@nocobase/client';
import React, { useContext } from 'react';
import { ChartV2BlockInitializer, ChartV2BlockDesigner, ChartInitializers, ChartV2Block } from './block';
import { ChartRenderer, G2PlotLibrary, ChartRendererProvider } from './renderer';
import { ChartLibraryProvider } from './renderer/ChartLibrary';

const Chart: React.FC = (props) => {
  const initializers = useContext(SchemaInitializerContext);
  const children = initializers.BlockInitializers.items[0].children;
  const has = children.some((initializer) => initializer.title === 'ChartV2');
  if (!has) {
    children.push({
      key: 'chart-v2',
      type: 'item',
      title: 'ChartV2',
      component: 'ChartV2BlockInitializer',
    });
  }
  return (
    <SchemaComponentOptions
      components={{ ChartV2BlockInitializer, ChartRenderer, ChartV2BlockDesigner, ChartV2Block, ChartRendererProvider }}
    >
      <SchemaInitializerProvider initializers={{ ...initializers, ChartInitializers }}>
        <ChartLibraryProvider name="G2Plot" charts={G2PlotLibrary}>
          {props.children}
        </ChartLibraryProvider>
      </SchemaInitializerProvider>
    </SchemaComponentOptions>
  );
};

export default Chart;
