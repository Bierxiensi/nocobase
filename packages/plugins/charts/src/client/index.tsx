import { registerValidateRules } from '@formily/core';
import { SchemaComponentOptions, SchemaInitializerContext, SettingsCenterProvider } from '@nocobase/client';
import JSON5 from 'json5';
import React, { useContext } from 'react';
import { ChartBlockEngine } from './ChartBlockEngine';
import { ChartBlockInitializer } from './ChartBlockInitializer';
import { ChartQueryMetadataProvider } from './ChartQueryMetadataProvider';
import './Icons';
import { lang } from './locale';
import { CustomSelect } from './select';
import { QueriesTable } from './settings/QueriesTable';

registerValidateRules({
  json5: (value, rule) => {
    if (!value) {
      return '';
    }
    try {
      const val = JSON5.parse(value);
      if (!isNaN(val)) {
        return {
          type: 'error',
          message: lang('Invalid JSON format'),
        };
      }
      return '';
    } catch (error) {
      console.error(error);
      return {
        type: 'error',
        message: lang('Invalid JSON format'),
      };
    }
  },
});

export default React.memo((props) => {
  const items = useContext(SchemaInitializerContext);
  const children = items.BlockInitializers.items[0].children;
  children.push({
    key: 'chart',
    type: 'item',
    icon: 'PieChartOutlined',
    title: '{{t("Chart",{ns:"charts"})}}',
    component: 'ChartBlockInitializer',
  });
  return (
    <ChartQueryMetadataProvider>
      <SettingsCenterProvider
        settings={{
          charts: {
            title: '{{t("Chart",{ns:"charts"})}}',
            icon: 'PieChartOutlined',
            tabs: {
              queries: {
                title: '{{t("Queries",{ns:"charts"})}}',
                component: () => <QueriesTable />,
              },
            },
          },
        }}
      >
        <SchemaComponentOptions components={{ CustomSelect, ChartBlockInitializer, ChartBlockEngine }}>
          <SchemaInitializerContext.Provider value={items}>{props.children}</SchemaInitializerContext.Provider>
        </SchemaComponentOptions>
      </SettingsCenterProvider>
    </ChartQueryMetadataProvider>
  );
});
