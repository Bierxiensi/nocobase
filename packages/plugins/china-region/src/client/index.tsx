import { Plugin } from '@nocobase/client';
import { ChinaRegionProvider } from './ChinaRegionProvider';

export class ChinaRegionPlugin extends Plugin {
  async load() {
    this.app.use(ChinaRegionProvider);
  }
}

export default ChinaRegionPlugin;
