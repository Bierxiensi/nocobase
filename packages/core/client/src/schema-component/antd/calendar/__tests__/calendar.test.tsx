import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import { CalendarBlockProvider } from '../../../../block-provider';
import { SchemaComponent, SchemaComponentProvider } from '../../../../schema-component';
import { ISchema } from '@formily/react';
import { AntdSchemaComponentProvider, APIClient, SchemaInitializerProvider } from '@nocobase/client';
import { CalendarV2 } from '../index';
import defaultValues from '../demos/defaultValues';

const schema = {
  type: 'array',
  name: 'calendar1',
  'x-component': 'CalendarV2',
  'x-component-props': {},
  default: defaultValues,
  properties: {
    toolBar: {
      type: 'void',
      'x-component': 'CalendarV2.ActionBar',
      properties: {
        today: {
          type: 'void',
          title: '今天',
          'x-component': 'CalendarV2.Today',
          'x-action': 'calendar:today',
          'x-align': 'left',
        },
        nav: {
          type: 'void',
          title: '翻页',
          'x-component': 'CalendarV2.Nav',
          'x-action': 'calendar:nav',
          'x-align': 'left',
        },
        title: {
          type: 'void',
          title: '标题',
          'x-component': 'CalendarV2.Title',
          'x-action': 'calendar:title',
          'x-align': 'left',
        },
        viewSelect: {
          type: 'void',
          title: '视图切换',
          'x-component': 'CalendarV2.ViewSelect',
          'x-action': 'calendar:viewSelect',
          'x-align': 'right',
        },
      },
    },
  },
};

const App= () => {
  return (
    <SchemaComponentProvider components={{CalendarV2,CalendarBlockProvider}}>
      <SchemaInitializerProvider>
        <AntdSchemaComponentProvider>
          <SchemaComponent schema={schema} />
        </AntdSchemaComponentProvider>
      </SchemaInitializerProvider>
    </SchemaComponentProvider>
  );
};

describe('calendar', () => {
  it('calendar', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
