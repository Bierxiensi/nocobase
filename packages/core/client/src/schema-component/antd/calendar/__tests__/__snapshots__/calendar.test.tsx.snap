// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar calendar 1`] = `
<div>
  <div
    style="height: 700px;"
  >
    <div
      class="rbc-calendar"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: center; overflow-x: auto; flex-shrink: 0;"
      >
        <div
          class="css-2ylt3q"
          style="display: flex; justify-content: space-between; align-items: center; width: 100%;"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <button
                class="ant-btn ant-btn-default"
                type="button"
              >
                <span>
                  Today
                </span>
              </button>
            </div>
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <div
                class="ant-btn-group"
              >
                <button
                  class="ant-btn ant-btn-default ant-btn-icon-only"
                  type="button"
                >
                  <span
                    aria-label="left"
                    class="anticon anticon-left"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="left"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                      />
                    </svg>
                  </span>
                </button>
                <button
                  class="ant-btn ant-btn-default ant-btn-icon-only"
                  type="button"
                >
                  <span
                    aria-label="right"
                    class="anticon anticon-right"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="right"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
            <div
              class="ant-space-item"
            >
              <div
                class="ant-btn-group"
                style="font-size: 1.75em; font-weight: 300;"
              >
                <span>
                  2022-12
                </span>
                <span
                  style="margin-left: 4px;"
                />
              </div>
            </div>
          </div>
          <div
            class="ant-space ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
            >
              <div
                class="ant-btn-group"
              >
                <div
                  class="ant-select ant-select-single ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                        aria-autocomplete="list"
                        aria-controls="rc_select_TEST_OR_SSR_list"
                        aria-haspopup="listbox"
                        aria-owns="rc_select_TEST_OR_SSR_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="rc_select_TEST_OR_SSR"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-item"
                      title="Month"
                    >
                      Month
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-label="Month View"
        class="rbc-month-view"
        role="table"
      >
        <div
          class="rbc-row rbc-month-header"
          role="row"
        >
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Sun
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Mon
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Tue
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Wed
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Thu
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Fri
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Sat
            </span>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    27
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    28
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    29
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    30
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    01
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    02
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    03
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    04
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    05
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    06
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    07
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    08
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    09
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    10
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    11
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    12
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    13
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    14
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    15
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    16
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    17
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    18
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    19
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    20
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    21
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    22
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    23
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    24
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg rbc-today"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    25
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    26
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    27
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    28
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell rbc-now rbc-current"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    29
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    30
                  </span>
                </a>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <a
                  role="cell"
                >
                  <span
                    class="rbc-date-solar"
                  >
                    31
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
